Code/Tools/FBuild/FBuildCore/BFF/Functions/FunctionPrint.cpp

```diff
@@ -169,27 +227,79 @@ FunctionPrint::FunctionPrint()
+                }
+                FLOG_BUILD( "%s]\n", indentStr.Get() );
             }
-            FLOG_BUILD( "%s]\n", indentStr.Get() );
             break;
         }
         case BFFVariable::VAR_ARRAY_OF_STRUCTS:
         {
             const auto & structs = var.GetArrayOfStructs();
-            FLOG_BUILD( "%s = // ArrayOfStructs, size: %u\n%s{\n", var.GetName().Get(), (uint32_t)structs.GetSize(), indentStr.Get() );
-            for ( const BFFVariable * subVar : structs )
+            FLOG_BUILD( "%s = // ArrayOfStructs, size: %u\n", var.GetName().Get(), (uint32_t)structs.GetSize() );
+            if ( maxIndent >= indent )
             {
-                PrintVarRecurse( *subVar, indent );
+                FLOG_BUILD( "%s{\n", indentStr.Get() );
+                for ( const BFFVariable * subVar : structs )
+                {
+                    PrintVarRecurse( *subVar, indent, maxIndent );
+                }
+                FLOG_BUILD( "%s}\n", indentStr.Get() );
             }
-            FLOG_BUILD( "%s}\n", indentStr.Get() );
             break;
         }
         case BFFVariable::MAX_VAR_TYPES: ASSERT( false ); break; // Something is terribly wrong
     }
 }
 
+
+// PrintVarRecurse
+//------------------------------------------------------------------------------
+/*static*/ void FunctionPrint::PrintStackVars( const AString& pattern, bool iterateParents, bool shouldRecurse )
+{
+    // Build an array of BFFStackFrame so it can be iterated in reverse order as 
+    // the frame stack (global variables get printed first)
+    Array< const BFFStackFrame* > aFrames(/*initialCapacity=*/10, /*resizable=*/true );
+    {
+        BFFStackFrame *pFrame = BFFStackFrame::GetCurrent();
+        if ( pFrame->GetParent() )
+        {
+            // Print gets a stack frame even though it doesn't have {}, and
+            // that frame isn't interesting to the caller.  So iterate past it.
+            pFrame = pFrame->GetParent();
+        }
+        do
+        {
+            aFrames.Append( pFrame );
+            pFrame = pFrame->GetParent();
+        } while ( pFrame != nullptr && iterateParents );
+    }
+    const uint32_t iNumFrames = static_cast<uint32_t>(aFrames.GetSize());
+
+    const uint32_t iMaxDepth = shouldRecurse ? 0xFFFFFFFF : 0;
+    const bool bPatternMatchesAll = (pattern == "*");
+    for ( uint32_t iFrame = iNumFrames; iFrame > 0; )
+    {
+        --iFrame;
+        const BFFStackFrame* const pFrame = aFrames[ iFrame ];
+        const uint32_t iDepth = iNumFrames - iFrame - 1;
+        const Array< const BFFVariable* >& rVariablesInFrame = pFrame->GetLocalVariables();
+        for ( size_t iVar = 0; iVar < rVariablesInFrame.GetSize(); ++iVar )
+        {
+            const BFFVariable* const pVar = rVariablesInFrame[ iVar ];
+            if ( bPatternMatchesAll || pVar->GetName().Matches( pattern.Get() ) )
+            {
+                PrintVarRecurse(*pVar, iDepth, iMaxDepth);
```

Comment:
The existence of this helper function is one of the reasons I opted to make this a power of Print instead of adding a new function